### PR TITLE
[OptionsResolver] Support union of types

### PR DESCRIPTION
**Pull request by VincentLanglet**
| Q             | A
| ------------- | ---
| Branch?       | 7.3
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | Fix #... 
| License       | MIT

Add support for `|` syntax when describing allowedType in OptionResolver.

It's not really useful for `int|string` since we can pass an array `['string', 'int']`.
But it's useful for array values since `(int|string)[]` was not possible so far and `['string[]', 'int[]']` was not the same thing.

### CODE REVIEW

File: src/Symfony/Component/OptionsResolver/Tests/OptionsResolverTest.php [hunk 1]
```diff
@@ -778,6 +778,33 @@ public function testSetAllowedTypesFailsIfUnknownOption()
         $this->resolver->setAllowedTypes('foo', 'string');
     }
 
+    public function testResolveTypedWithUnion()
+    {
+        $this->resolver->setDefined('foo');
+        $this->resolver->setAllowedTypes('foo', 'string|int');
+        $options = $this->resolver->resolve(['foo' => 1]);
+
+        $this->assertSame(['foo' => 1], $options);
```
Comment (yceruto) [inline src/Symfony/Component/OptionsResolver/Tests/OptionsResolverTest.php:787]: please add another assert for a `string` value to complete this test case

File: src/Symfony/Component/OptionsResolver/Tests/OptionsResolverTest.php [hunk 2]
```diff
@@ -778,6 +778,33 @@ public function testSetAllowedTypesFailsIfUnknownOption()
         $this->resolver->setAllowedTypes('foo', 'string');
     }
 
+    public function testResolveTypedWithUnion()
+    {
+        $this->resolver->setDefined('foo');
+        $this->resolver->setAllowedTypes('foo', 'string|int');
```
Comment (yceruto) [inline src/Symfony/Component/OptionsResolver/Tests/OptionsResolverTest.php:784]: it's the same as `->setAllowedTypes('foo', ['string', 'int']);` right? I like it

File: src/Symfony/Component/OptionsResolver/Tests/OptionsResolverTest.php [hunk 3]
```diff
@@ -778,6 +778,33 @@ public function testSetAllowedTypesFailsIfUnknownOption()
...
+        $this->resolver->setAllowedTypes('foo', '(string)|(int)');
+        $options = $this->resolver->resolve(['foo' => 1]);
+
+        $this->assertSame(['foo' => 1], $options);
+    }
+
+    public function testResolveTypedWithUnionUselessParentheses2()
+    {
+        $this->resolver->setDefined('foo');
+        $this->resolver->setAllowedTypes('foo', '((string)|(int))');
```
Comment (yceruto) [inline src/Symfony/Component/OptionsResolver/Tests/OptionsResolverTest.php]: If useless, I'd prefer to throw an exception (invalid type syntax) instead, as this could be a case where you might be missing the `[]` at the end. I believe an exception that forces you to remove the unnecessary parentheses would be more helpful than allowing them for no reason.

**Review by yceruto — CHANGES_REQUESTED** [review]
Good addition 👍 thanks!

File: src/Symfony/Component/OptionsResolver/Tests/OptionsResolverTest.php (hunk 2 above)
Comment (VincentLanglet) [reply src/Symfony/Component/OptionsResolver/Tests/OptionsResolverTest.php:784]: Yes it's the same

**Comment by yceruto** [comment]
Also, we could find a regex that matches all possible variants instead of checking them one by one.

**Comment by VincentLanglet** [comment]
> Also, we could find a regex that matches all possible variants instead of checking them one by one.

Didn't find an easy one, but I'm opened to suggestion.

I preferred to add the logic in the existing recursive implementation

File: src/Symfony/Component/OptionsResolver/Tests/OptionsResolverTest.php (hunk 3 above)
Comment (VincentLanglet) [reply src/Symfony/Component/OptionsResolver/Tests/OptionsResolverTest.php]: What do you consider as useless parentheses ?
- (A) ?
- (A|B) ? But since the called was made as recursive it might come from `(A|B)[]`
- (A|B)|(C|D) ? But this doesn't seems easy to report this as useless since
  - It's split as `(A|B)` and `(C|D)` and then we're in the previous case

Since the call is recursive it doesn't seems easy to report useless parenthesis or maybe it will require an extra param to know if we just removed `[]`.

Also, since the validation is made in Runtime (when calling `resolve`) it seems not as useful as it would have been when calling `setAllowedTypes`...

**Comment by VincentLanglet** [comment]
I found edge case where my previous implementation remove badly parenthesis, so I now introduce a private method `isEnclosedWithParenthesis` @yceruto

File: src/Symfony/Component/OptionsResolver/Tests/OptionsResolverTest.php (hunk 3 above)
Comment (VincentLanglet) [reply src/Symfony/Component/OptionsResolver/Tests/OptionsResolverTest.php]: Also, there is so far no similar validation about allowedTypes passed.
- `string[` is not reported even if it's obviously a mistake
- `['string', 'string']` is not reported
- `strin[]` is not reported

I think introducing an InvalidType exceptions for those syntax (and useless parenthesis if wanted) might be better in another PR.

Comment (yceruto) [reply src/Symfony/Component/OptionsResolver/Tests/OptionsResolverTest.php]: Agree, but let’s not add test cases for them though

Comment (VincentLanglet) [reply src/Symfony/Component/OptionsResolver/Tests/OptionsResolverTest.php]: Done

File: src/Symfony/Component/OptionsResolver/Tests/OptionsResolverTest.php [hunk 4]
```diff
@@ -778,6 +778,31 @@ public function testSetAllowedTypesFailsIfUnknownOption()
...
+
+        $options = $this->resolver->resolve(['foo' => '1']);
+        $this->assertSame(['foo' => '1'], $options);
+    }
+
+
+    public function testResolveTypedWithUnionOfArray()
+    {
+        $this->resolver->setDefined('foo');
+        $this->resolver->setAllowedTypes('foo', '((string|int)[])|((bool|int)[])');
```
Comment (yceruto) [inline src/Symfony/Component/OptionsResolver/Tests/OptionsResolverTest.php]: Isn't it the same as `(string|int|bool)[]`?

Comment (VincentLanglet) [reply src/Symfony/Component/OptionsResolver/Tests/OptionsResolverTest.php]: Technically no.
`[1, 'foo', bool]` is accepted by `(string|int|bool)[]` but not by `((string|int)[])|((bool|int)[])` since the array is not accepted by `(string|int)[]` or `(bool|int)[]`.

This example is not useful since I don't expect someone to really use this exact definition, but I wanted to write a test with some complex expression `(SomeExpressionWithParenthesis|AnotherExpressionWithParenthesis)`, because my first implementation of removing `()`
- changed `((string|int)[])|((bool|int)[])` to `(string|int)[])|((bool|int)[]`
- consider as the array of `(string|int)[])|((bool|int)`
- then, checked either `string|int)[]` or `(bool|int`

So this is mainly a test for the non-enclosing parenthesis

Comment (yceruto) [reply src/Symfony/Component/OptionsResolver/Tests/OptionsResolverTest.php]: I see, so you want an array of `string|int` or an array of `bool|int` but  not a mix of them.

Comment (yceruto) [reply src/Symfony/Component/OptionsResolver/Tests/OptionsResolverTest.php]: `((string|int)[])|((bool|int)[])` why the extra parentheses are necessary in this case? can't it be just `(string|int)[]|(bool|int)[]`?

**Comment by yceruto** [comment]
Could you confirm that class Union is also supported with this new syntax? i.e. `(App\FooInterface|App\BarInterface)[]`

File: src/Symfony/Component/OptionsResolver/Tests/OptionsResolverTest.php (hunk 4 above)
Comment (VincentLanglet) [reply src/Symfony/Component/OptionsResolver/Tests/OptionsResolverTest.php]: I had an issue about the order the string was processed.

`int|(string|bool)[] was considered as an array of `int|(string|bool)` rather than an union of `int` and `(string|bool)[]`.

But I reworked the implementation to split on `|` before removing `[]` and I think I finally found a way to give `|` a better priority (to use the same syntax than in phpdoc).

**Comment by VincentLanglet** [comment]
> Could you confirm that class Union is also supported with this new syntax? i.e. `(App\FooInterface|App\BarInterface)[]`

I added a test for `\DateTime::class.'|'.\DateTimeImmutable::class`. Is it enough ?

**Review by nicolas-grekas — APPROVED** [review]

**Review by yceruto — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @VincentLanglet.
